ale_linters/nix/nix.vim

```diff
@@ -1,18 +1,51 @@
 " Author: Alistair Bill <@alibabzo>
+" Author: Maximilian Bosch <maximilian@mbosch.me>
 " Description: nix-instantiate linter for nix files
 
+if !exists('g:ale_linters_nix_version')
+    let g:ale_linters_nix_extra_cmd = ''
+
+    " strip away `nix-instantiate (Nix) `
+    let g:version = system('nix-instantiate --version')[22:]
```

Comment:
This `system()` method is blocking so cannot be merged into ALE. I got PR's [rejected](https://github.com/dense-analysis/ale/pull/2121#discussion_r239214535) by @w0rp time ago for the same reason.

The way to do this now is by chaining commands using `ale#command#Run()` function. Read the help of this function and look into `semver.vim` file the function `ale#semver#RunWithVersionCheck()` for an example of how to achieve what you want.

There are are many other linters/fixers that check the tool version using the same chained command method. Grepping ALE's code for WithVersion will let you find them.

Reply:
Hmm the problem is that this doesn't control which command to execute, but how to interpret the output (in contrast to the other linters AFAIU). Do you have a suggestion how to implement this with `RunWithVersionCheck`?

Reply:
You can use ale#command#Run() to build the command line including or not the ` ' --log-format internal-json '` flag depending on the tool version.

For the Handle() method you first parse using the pattern and check if l:output is empty or not. If not empty return it and if empty try parsing with the json format:

```
function! ale_linters#nix#nix#Handle(buffer, lines) abort
    let l:pattern = '^\(.\+\): \(.\+\), at .*:\(\d\+\):\(\d\+\)$'
    let l:output = []

    for l:match in ale#util#GetMatches(a:lines, l:pattern)
        call add(l:output, {
        \   'lnum': l:match[3] + 0,
        \   'col': l:match[4] + 0,
        \   'text': l:match[1] . ': ' . l:match[2],
        \   'type': l:match[1] =~# '^error' ? 'E' : 'W',
        \})
    endfor

    if !empty(l:output)
        return l:output
    endif

    for l:line in a:lines
        if l:line =~# '^@nix .*'
            let l:result = json_decode(strpart(l:line, 4))

            if has_key(l:result, 'column')
                call add(l:output, {
                \     'type': 'E',
                \     'lnum': l:result.line,
                \     'col': l:result.column,
                \     'text': l:result.raw_msg
                \})
            endif
        endif
     endfor
   
    return l:output
endfunction
```

Reply:
Fixed, thanks for the hint :)